File: 2022/09/03/SingleNumber.swift
/**
 https://leetcode.com/problems/single-number/
 - Author: Andrew Jarombek
 - Date: 9/3/2022
 */

class Solution {
    func singleNumber(_ nums: [Int]) -> Int {
        var items: Set<Int> = []

        for num in nums {
            if items.contains(num) {
                items.remove(num)
            } else {
                items.insert(num)
            }
        }

        for item in items {
            return item
        }

        return 0
    }
}

let solution = Solution()
let result = solution.singleNumber([2, 2, 1])
print(result)

File: 2022/11/19/cpp/singleNumber.cpp
// https://leetcode.com/problems/single-number/
// Author: Andrew Jarombek
// Date: 11/19/2022

#include <cassert>
#include <vector>
#include <set>

using namespace std;

int singleNumber(vector<int>& nums) {
    set<int> s {};

    for (int num : nums) {
        if (s.find(num) != s.end()) {
            s.erase(num);
        } else {
            s.insert(num);
        }
    }

    for (int num : s) {
        return num;
    }

    return -1;
}

int main() {
    vector<int> v {2, 2, 1};
    assert(singleNumber(v) == 1);
}